### PR TITLE
Make `LogicalPlan::with_new_exprs,` deprecate `from_plan`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?
related to https://github.com/apache/arrow-datafusion/pull/7359

## Rationale for this change

While reviewing a great documentation improvement from  @tshauck  it took me a while to find the `from_plan` function used to rewrite LogicalPlans. See https://github.com/apache/arrow-datafusion/pull/7359#discussion_r1304408534

I think this makes it harder to work with the DataFusion codebase unnecessarily.

The reason it took me a while to find I think was
1.  Isn't method on `LogicalPlan` (it is in a utility module). 
2. It didn't have `expr` or `exprs` in its name

Thus, given it takes a `&LogicalPlan` as its first input I think it makes sense and would be more easily findable as a method on `LogicalPlan`.

It also turns out `from_plan` forces `Expr`s to be copied even when most callsites already have a copy of those exprs

## What changes are included in this PR?

1. Move from `from_plan` to  `LogicalPlan::with_new_exprs()` 
4. deprecate `from_plan`
5. Avoid a copy in `new_with_exprs` (see second commit in this PR)

Note that `LogicalPlan::new_with_exprs` take a `Vec<Expr>` instead of `&[Expr]` like `from_plan` as it needs the owned Exprs anyways

## Are these changes tested?
By existing tests

## Are there any user-facing changes?
New API, deprecated old one

### CODE REVIEW

**Review by tshauck — APPROVED** [review]
This seems like a nice refactor from my point of view... it would've helped me find the functionality initially.

It'd also be nice to merge this (or decide not to) so the library docs can be kept correct relative to the underlying code.

**Comment by alamb** [comment]
@liukun4515 / @jackwener could I trouble you for a review of this PR? I can no longer merge such PRs without an approval of another committer. 

Thank you, Andrew

**Review by jackwener — APPROVED** [review]
A nice change to me, thanks @alamb

**Pull request merged**
